lib/plugins/aws/package/compile/functions/index.js

```diff
@@ -441,98 +441,125 @@ class AwsCompileFunctions {
+        versionCompilationPromise = this.addFileContentsToHashes(artifactFilePath, [
+          fileHash,
+          versionHash,
+        ])
+          .then(() => {
+            // Include all referenced layer code in the version id hash
+            const functionProperties = _.omit(_.get(functionResource, 'Properties', {}), 'Code');
+            const layerProperties = this.extractLayerPropertiesFromFunction(
+              functionProperties,
+              cfTemplate
+            );
+
+            const layerZipFilePaths = [];
+            layerProperties.forEach(layer => {
+              const layerZipFileName = layer.properties.Properties.Content.S3Key.split(
+                path.sep
+              ).pop();
+              const layerZipFilePath = path.join(this.packagePath, layerZipFileName);
+              layerZipFilePaths.push(layerZipFilePath);
             });
-        }).then(() => {
-          // Include function configuration in version id hash (without the Code part)
-          const properties = _.omit(_.get(functionResource, 'Properties', {}), 'Code');
-          _.forOwn(properties, value => {
-            if (_.isObject(value)) {
-              versionHash.write(JSON.stringify(value));
-            } else {
-              versionHash.write(value != null ? String(value) : '');
-            }
-          });
 
-          // Finalize hashes
-          fileHash.end();
-          versionHash.end();
+            return BbPromise.map(layerZipFilePaths, zipFileName => {
+              return this.addFileContentsToHashes(zipFileName, [versionHash]);
+            });
+          })
+          .then(() => {
+            // Include function and layer configuration in version id hash (without the Code part)
+            const functionProperties = _.omit(_.get(functionResource, 'Properties', {}), 'Code');
+            const layerProperties = this.extractLayerPropertiesFromFunction(
+              functionProperties,
+              cfTemplate
+            );
 
-          const fileDigest = fileHash.read();
-          const versionDigest = versionHash.read();
+            const mappedLayerFunctionProperties = layerProperties.map(layer => {
+              const clonedLayer = _.cloneDeep(layer);
+              clonedLayer.properties.Properties.Content.S3Key = '';
+              return clonedLayer.properties;
+            });
 
-          const versionResource = this.cfLambdaVersionTemplate();
+            mappedLayerFunctionProperties.push(functionProperties);
+
+            mappedLayerFunctionProperties.forEach(property => {
+              Object.keys(property).forEach(key => {
+                const value = property[key];
+                if (_.isObject(value)) {
+                  versionHash.write(JSON.stringify(value));
+                } else {
+                  versionHash.write(value != null ? String(value) : '');
+                }
+              });
+            });
```

Comment:
It's a good opportunity, to further improve this logic.

Currently in master we take into account only property values, which is error prone, as replacing `foo: 1` with `bar: 1` in function config, will not incur hash change, while such change should be recognized.

Additionally this is not bulletproof for case where properties stay same but are put in other order.

We can fix it and simplify the logic, with:

```javascript
const byKey = ([key1], [key2]) => key1.localeCompare(key2)
functionProperties.layers = {}
for (const { name, properties: layerProperties } of layerConfigurations) {
  functionProperties.layers[name] = _.fromPairs(_.entries(layerProperties).sort(byKey))
}

hash.write(JSON.stringify(_.fromPairs(_.entries(functionProperties).sort(byKey))))
```